File: July-LeetCoding-Challenge-2020/Week1/107-Binary-Tree-Level-Order-Traversal-II.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

// Solution #1 Breadth First Search
// TC: O(n) number of nodes
// SC: O(w) max size of width of tree
// Runtime: 4 ms, faster than 92.73%
// Memory Usage: 12.8 MB, less than 59.32%
class Solution {
    
    vector<vector<int>> res;
    stack<vector<int>> s;
    
public:
    
    vector<vector<int>> levelOrderBottom(TreeNode* root) {
    
        if(!root)
            return res;
        
        BFS(root);
        
        return getResult();
    }
    
    void BFS(TreeNode *root){
        
        int qSize;
        queue<TreeNode*> q;
        vector<int> valueByLevel;

        q.push(root);
        
        while(!q.empty()){
            
            qSize = q.size();
            
            for(int i = 0; i < qSize; i++){
                TreeNode *node = q.front();
                q.pop();
                
                valueByLevel.push_back(node->val);
                
                if(node->left)
                    q.push(node->left);
                
                if(node->right)
                    q.push(node->right);
            }
            
            s.push(valueByLevel);
            valueByLevel.clear();
        }
    }
    
    vector<vector<int>> getResult(){
        
        while(!s.empty()){
            
            res.push_back(s.top());
            s.pop();
        }
        
        return res;
    }
};

File: July-LeetCoding-Challenge-2020/Week1/107-binary-tree-level-order-traversal-ii.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    # Sol 1. DFS    
    def levelOrderBottom(self, root: TreeNode) -> List[List[int]]:
        result=[]
        if not root:
            return result
        self.dfs(root,0,result)
        
        return result
    
    def dfs(self,root,level,res):
        if not root:
            return 
        if len(res) < level+1: # current level
            res.append([])
        self.dfs(root.left, level+1, res)
        self.dfs(root.right, level+1, res)
        res[level].append(root.val)
    # TC: O(N)
    # SC: O(N)

    # Sol 2. BFS
    def levelOrderBottom(self, root: TreeNode) -> List[List[int]]:
        levels = []
        next_level = deque([root])
        
        while root and next_level:
            curr_level = next_level
            next_level = deque()
            levels.append([])
            
            for node in curr_level:
                # append the current node value
                levels[-1].append(node.val)
                # process child nodes for the next level
                if node.left:
                    next_level.append(node.left)
                if node.right:
                    next_level.append(node.right)
                    
        return levels[::-1]
        # TC: O(N)
        # SC: O(N)